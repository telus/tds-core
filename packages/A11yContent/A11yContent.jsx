import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { safeRest } from '@tds/util-helpers'
import { componentWithName, or } from '@tds/util-prop-types'

const StyledA11yContent = styled.span({
  position: 'absolute',
  height: '1px',
  width: '1px',
  overflow: 'hidden',
  clip: 'rect(1px, 1px, 1px, 1px)',
})

/**
 * @version ./package.json
 */

const A11yContent = ({ children, ...rest }) => {
  return <StyledA11yContent {...safeRest(rest)}>{children}</StyledA11yContent>
}

A11yContent.propTypes = {
  /**
   * Accessible content. Can be either a string or a `Heading` Component
   */
  children: or([PropTypes.string, componentWithName('Heading')]).isRequired,
}
A11yContent.defaultProps = {}

export default A11yContent
