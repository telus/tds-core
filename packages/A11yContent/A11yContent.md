Allows content to be hidden from sight, but still visible to screen readers. This can be used to give context to users using screen readers without revealing unnecessary extra text to sighted users.

```jsx
    <Text>The screen reader says:</Text><A11yContent>This is accessible content</A11yContent>
```

If `A11yContent` has to go between two separate text components, considering wrapping the whole text in a `<span>` to persist leading and trailing space.

```jsx
<ButtonLink href="#">
  <span>
    Shop <A11yContent>iPhone</A11yContent> Now
  </span>
</ButtonLink>
```
