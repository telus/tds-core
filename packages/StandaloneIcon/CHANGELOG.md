# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 2.2.3 (2022-02-16)


### Bug Fixes

* gha release job npm publish token ([ce7aacd](https://github.com/telusdigital/tds/commit/ce7aacdd77e54c4f8e48eb31d0c05dc804530324))





## 2.2.2 (2022-02-15)


### Bug Fixes

* gha release job git config ([276ebb6](https://github.com/telusdigital/tds/commit/276ebb6968a0b56c9b87c178f6895a60ae108e71))





## 2.2.1 (2022-02-15)


### Bug Fixes

* gha release job git config ([1ab603d](https://github.com/telusdigital/tds/commit/1ab603d68c36219b0711fc353bc2515b64712ca9))





# [2.2.0](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.19...@tds/core-standalone-icon@2.2.0) (2021-09-21)


### Features

* **core-standalone-icon:** add react 17 to peer deps ([a2442b1](https://github.com/telusdigital/tds/commit/a2442b1d871057225de8037cc3d6be6f22fd4b1e))





## [2.1.19](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.18...@tds/core-standalone-icon@2.1.19) (2021-01-15)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.18](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.17...@tds/core-standalone-icon@2.1.18) (2020-09-11)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.17](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.16...@tds/core-standalone-icon@2.1.17) (2020-09-08)


### Bug Fixes

* **core-standalone-icon:** add support for styled-components v5 ([b98c35c](https://github.com/telusdigital/tds/commit/b98c35cfa404b2132a54f6a957d963594d260f0b))





## [2.1.16](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.15...@tds/core-standalone-icon@2.1.16) (2020-09-04)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.15](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.14...@tds/core-standalone-icon@2.1.15) (2020-08-18)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.14](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.13...@tds/core-standalone-icon@2.1.14) (2020-08-07)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.13](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.12...@tds/core-standalone-icon@2.1.13) (2020-02-03)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.12](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.11...@tds/core-standalone-icon@2.1.12) (2020-01-27)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.11](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.10...@tds/core-standalone-icon@2.1.11) (2020-01-23)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.10](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.9...@tds/core-standalone-icon@2.1.10) (2020-01-22)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.9](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.8...@tds/core-standalone-icon@2.1.9) (2019-11-20)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.8](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.7...@tds/core-standalone-icon@2.1.8) (2019-10-09)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.7](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.6...@tds/core-standalone-icon@2.1.7) (2019-10-04)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.6](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.5...@tds/core-standalone-icon@2.1.6) (2019-09-12)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.5](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.4...@tds/core-standalone-icon@2.1.5) (2019-09-06)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.4](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.3...@tds/core-standalone-icon@2.1.4) (2019-08-26)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.3](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.2...@tds/core-standalone-icon@2.1.3) (2019-08-22)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.2](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.1...@tds/core-standalone-icon@2.1.2) (2019-08-09)

**Note:** Version bump only for package @tds/core-standalone-icon





## [2.1.1](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.1.0...@tds/core-standalone-icon@2.1.1) (2019-08-01)

**Note:** Version bump only for package @tds/core-standalone-icon





# [2.1.0](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.0.1...@tds/core-standalone-icon@2.1.0) (2019-07-31)


### Features

* **core-standalone-icon:** add 20px and 32px sizes ([8fb1f71](https://github.com/telusdigital/tds/commit/8fb1f71))





## [2.0.1](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@2.0.0...@tds/core-standalone-icon@2.0.1) (2019-07-30)

**Note:** Version bump only for package @tds/core-standalone-icon





# [2.0.0](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.1.2...@tds/core-standalone-icon@2.0.0) (2019-07-11)


### Features

* **core-standalone-icon:** upgrade to styled-components ([31bd31c](https://github.com/telusdigital/tds/commit/31bd31c))


### BREAKING CHANGES

* **core-standalone-icon:** upgrade to styled-components





## [1.1.2](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.1.1...@tds/core-standalone-icon@1.1.2) (2019-07-10)


### Bug Fixes

* **core-standalone-icon:** remove aria-label attribute ([db871b7](https://github.com/telusdigital/tds/commit/db871b7))





## [1.1.1](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.1.0...@tds/core-standalone-icon@1.1.1) (2019-06-26)


### Bug Fixes

* **core-standalone-icon:** render a11yText in A11yContent ([1d0241e](https://github.com/telusdigital/tds/commit/1d0241e)), closes [#985](https://github.com/telusdigital/tds/issues/985)





# [1.1.0](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.0.5...@tds/core-standalone-icon@1.1.0) (2019-06-04)

### Features

- **core-standalone-icon:** add innerRef prop to allow forwarding of ref ([fed0f6c](https://github.com/telusdigital/tds/commit/fed0f6c))

## [1.0.5](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.0.4...@tds/core-standalone-icon@1.0.5) (2019-05-01)

**Note:** Version bump only for package @tds/core-standalone-icon

## [1.0.4](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.0.3...@tds/core-standalone-icon@1.0.4) (2019-04-22)

**Note:** Version bump only for package @tds/core-standalone-icon

## [1.0.3](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.0.2...@tds/core-standalone-icon@1.0.3) (2019-03-18)

**Note:** Version bump only for package @tds/core-standalone-icon

## [1.0.2](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.0.1...@tds/core-standalone-icon@1.0.2) (2019-03-05)

**Note:** Version bump only for package @tds/core-standalone-icon

## [1.0.1](https://github.com/telusdigital/tds/compare/@tds/core-standalone-icon@1.0.0...@tds/core-standalone-icon@1.0.1) (2019-02-06)

**Note:** Version bump only for package @tds/core-standalone-icon
