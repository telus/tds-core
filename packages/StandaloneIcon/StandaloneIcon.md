### Making interactive icons

Use the `onClick` prop to create an accessible interactive icon.

Interactive icons will have a minimum click/tap area of about 32px, or 8-10mm, to ensure that they can be touched easily on
mobile devices. Take care not to place other elements too close to interactive icons, as the touch area could overlap.

```
<StandaloneIcon
  symbol="spyglass"
  size={24}
  a11yText="Search this site."
  onClick={() => console.log("searching...")}
/>
```
