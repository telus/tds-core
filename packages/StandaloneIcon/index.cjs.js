const StandaloneIcon = require('./dist/index.cjs')

module.exports = StandaloneIcon
