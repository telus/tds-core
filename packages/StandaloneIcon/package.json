{
  "name": "@tds/core-standalone-icon",
  "version": "1.1.2",
  "description": "An icon that has meaning within the context of the page",
  "main": "index.cjs.js",
  "module": "index.es.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/telusdigital/tds.git"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "TELUS digital",
  "engines": {
    "node": ">=8"
  },
  "bugs": {
    "url": "https://github.com/telusdigital/tds/issues"
  },
  "homepage": "http://tds.telus.com",
  "peerDependencies": {
    "react": "^16.8.2",
    "react-dom": "^16.8.2",
    "styled-components": "^4.1.3"
  },
  "dependencies": {
    "@tds/core-a11y-content": "^2.1.0",
    "@tds/shared-styles": "^1.3.0",
    "prop-types": "^15.5.10"
  },
  "devDependencies": {
    "@tds/core-colours": "^1.1.2"
  }
}
