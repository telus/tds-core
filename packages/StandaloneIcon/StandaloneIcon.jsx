import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import A11yContent from '@tds/core-a11y-content'
import { buttons } from '@tds/shared-styles'

import safeRest from '../../shared/utils/safeRest'

import Icon from '../../shared/components/Icon/Icon'

const mobileDeviceTapArea = 32 // https://www.w3.org/TR/mobile-accessibility-mapping/#touch-target-size-and-spacing
const touchAreaStyles = iconSize => {
  const length = (mobileDeviceTapArea - iconSize) / 2

  return {
    padding: `${length}px`,
    margin: `-${length}px`,
  }
}

const StandaloneIconClickable = styled.button(buttons.noStyle)

/**
 * An icon that has meaning within the context of the page, which should be communicated to screen readers.
 *
 * @version ./package.json
 */
const StandaloneIcon = ({ symbol, variant, size, onClick, a11yText, innerRef, ...rest }) => {
  const iconProps = {
    symbol,
    variant,
    size,
  }

  if (onClick) {
    const needsExpandedTouchArea = size < mobileDeviceTapArea

    return (
      <StandaloneIconClickable
        {...safeRest(rest)}
        ref={innerRef}
        onClick={onClick}
        style={needsExpandedTouchArea ? touchAreaStyles(size) : undefined}
      >
        <Icon {...iconProps}>
          <A11yContent>{a11yText}</A11yContent>
        </Icon>
      </StandaloneIconClickable>
    )
  }

  return <Icon {...safeRest(rest)} {...iconProps} />
}

StandaloneIcon.propTypes = {
  /**
   * Name of the icon symbol.
   */
  symbol: PropTypes.oneOf([
    'caretDown',
    'caretUp',
    'checkmark',
    'chevron',
    'leftChevron',
    'exclamationPointCircle',
    'expander',
    'hamburger',
    'location',
    'minus',
    'plus',
    'questionMarkCircle',
    'spyglass',
    'times',
  ]).isRequired,
  /**
   * The appearance of the Icon.
   */
  variant: PropTypes.oneOf(['primary', 'secondary', 'inverted', 'error']),
  /**
   * The icon size in pixels.
   */
  size: PropTypes.oneOf([16, 24, 48]),
  /**
   * Pass a handler to the icon to make it interactive. Wraps the icon with a `<button>`.
   */
  onClick: PropTypes.func,
  /**
   * A description of the icon for screen readers.
   */
  a11yText: PropTypes.string.isRequired,
  /* @ignore */
  innerRef: PropTypes.object,
}

StandaloneIcon.defaultProps = {
  variant: undefined,
  size: 24,
  onClick: undefined,
  innerRef: undefined,
}

export default StandaloneIcon
