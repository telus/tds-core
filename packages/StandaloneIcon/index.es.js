import StandaloneIcon from './dist/index.es'

export default StandaloneIcon
