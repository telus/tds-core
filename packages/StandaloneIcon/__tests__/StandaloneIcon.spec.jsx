import React from 'react'
import { shallow, render } from 'enzyme'

import A11yContent from '@tds/core-a11y-content'

import StandaloneIcon from '../StandaloneIcon'

describe('StandaloneIcon', () => {
  describe('Plain StandaloneIcon', () => {
    const defaultProps = {
      symbol: 'spyglass',
      a11yText: 'Some text for the screen readers.',
    }
    const doShallow = (props = {}) => shallow(<StandaloneIcon {...defaultProps} {...props} />)

    it('renders', () => {
      const icon = render(<StandaloneIcon {...defaultProps} />)
      expect(icon).toMatchSnapshot()
    })

    it('passes attributes to the Icon component', () => {
      const icon = doShallow({
        symbol: 'checkmark',
        variant: 'secondary',
        size: 16,
        id: 'the-icon',
      })

      expect(icon).toHaveProp('symbol', 'checkmark')
      expect(icon).toHaveProp('variant', 'secondary')
      expect(icon).toHaveProp('size', 16)
      expect(icon).toHaveProp('id', 'the-icon')
    })

    it('does not allow custom CSS', () => {
      const icon = doShallow({
        className: 'my-custom-class',
        style: { color: 'hotpink' },
      })

      expect(icon).not.toHaveProp('className', 'my-custom-class')
      expect(icon).not.toHaveProp('style')
    })
  })

  describe('Interactive icon', () => {
    const defaultProps = {
      symbol: 'spyglass',
      a11yText: 'Some text for the screen readers.',
      onClick: jest.fn(),
    }
    const doShallow = (props = {}) => {
      const interactiveIcon = shallow(<StandaloneIcon {...defaultProps} {...props} />)

      return {
        interactiveIcon,
        button: interactiveIcon.dive(),
      }
    }

    it('renders', () => {
      const interactiveIcon = render(<StandaloneIcon {...defaultProps} />)

      expect(interactiveIcon).toMatchSnapshot()
    })

    it('passes a11yText to the button as A11yContent', () => {
      const { interactiveIcon } = doShallow({ onClick: jest.fn() })

      expect(interactiveIcon.find('Icon')).toHaveProp(
        'children',
        <A11yContent>Some text for the screen readers.</A11yContent>
      )
    })

    it('triggers the click handler', () => {
      const onClickMock = jest.fn()
      const { interactiveIcon } = doShallow({ onClick: onClickMock })

      interactiveIcon.simulate('click')

      expect(onClickMock).toHaveBeenCalled()
    })

    describe('touch area', () => {
      it('expands the touch area to be large enough for a finger when the icon is small', () => {
        let { button } = doShallow({ size: 16 })
        expect(button).toHaveStyle('padding', '8px')
        expect(button).toHaveStyle('margin', '-8px')

        button = doShallow({ size: 24 }).button

        expect(button).toHaveStyle('padding', '4px')
        expect(button).toHaveStyle('margin', '-4px')
      })

      it('does not expand the touch area when the icon is large', () => {
        const { button } = doShallow({ size: 48 })

        expect(button).not.toHaveStyle('padding')
        expect(button).not.toHaveStyle('margin')
      })
    })

    it('passes additional attributes to the button element', () => {
      const { button } = doShallow({
        id: 'the-interactiveIcon',
        role: 'button',
      })

      expect(button).toHaveProp('id', 'the-interactiveIcon')
      expect(button).toHaveProp('role', 'button')
    })

    it('does not allow custom CSS', () => {
      const { button } = doShallow({
        className: 'my-custom-class',
        style: { color: 'hotpink' },
      })

      expect(button).not.toHaveProp('className', 'my-custom-class')
      expect(button).not.toHaveStyle('color', 'hotpink')
    })
  })
})
