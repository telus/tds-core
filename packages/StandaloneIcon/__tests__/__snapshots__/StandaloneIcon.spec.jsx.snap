// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandaloneIcon Interactive icon renders 1`] = `
.c2 {
  position: absolute;
  height: 1px;
  width: 1px;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c1 {
  font-family: 'TELUS Core Icons';
  display: inline-block;
  font-weight: normal;
  font-style: normal;
  speak: none;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  text-transform: none;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  line-height: 1;
  vertical-align: middle;
  font-size: 1.5rem;
}

.c1::before {
  content: '\\f111';
}

.c0 {
  padding: 0;
  margin: 0;
  border-width: 0;
  font-family: "TELUS-Web","Helvetica Neue",Helvetica,Arial,sans-serif;
  color: #2a2c2e;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  box-shadow: none;
  cursor: pointer;
}

<button
  class="c0"
  style="padding:4px;margin:-4px"
>
  <i
    aria-label="Some text for the screen readers."
    class="c1"
  >
    <span
      class="c2"
    >
      Some text for the screen readers.
    </span>
  </i>
</button>
`;

exports[`StandaloneIcon Plain StandaloneIcon renders 1`] = `
.c0 {
  font-family: 'TELUS Core Icons';
  display: inline-block;
  font-weight: normal;
  font-style: normal;
  speak: none;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  text-transform: none;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  line-height: 1;
  vertical-align: middle;
  font-size: 1.5rem;
}

.c0::before {
  content: '\\f111';
}

<i
  aria-label="Some text for the screen readers."
  class="c0"
/>
`;
